### PR TITLE
sys/net/gnrc/pktbuf_static: Fix alignment

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

The buffer _pktbuf was previously only aligned to a 1 byte boundary. This could result in run time issues. This PR makes sure the buffer is aligned as required and silences warnings generated by `-Wcast-align` where correct alignment is indeed ensured.

### Testing procedure

Hopefully, this doesn't change any binaries. At least the `.text` section should remain unchanged.

### Issues/PRs references

Split out of https://github.com/RIOT-OS/RIOT/pull/14955

### CODE REVIEW

File: sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c [hunk 1]
```diff
@@ -424,7 +428,9 @@ static void *_pktbuf_alloc(size_t size)
         }
     }
     else {
-        _unused_t *new = (_unused_t *)(((uint8_t *)ptr) + size);
+        /* alignment is ensured by rounding size up in the _align() function.
+         * We cast to uintptr_t as intermediate step to silence -Wcast-align */
+        _unused_t *new = (_unused_t *)((uintptr_t)ptr + size);
```
Comment (miri64) [inline sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:433]: Doesn't this result in an `size * sizeof(uintptr_t)` bytes offset? I assume `uintptr_t` to be word length.

**Review by miri64 — COMMENTED** [review]
Actually the 4 byte alignment is already achieved in code (even 8 bytes because we need to fit the (next-pointer, size)-pair into the "holes"... Unless of course the linker puts the array into RAM unaligned to begin with...

Anyway, not sure about the pointer arithmetic here.

File: sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c (hunk 1 above)
Comment (maribu) [reply sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:433]: `uintptr_t` is an pointer sized integer. So you have integer arithmetic here, not pointer arithmetic.

File: sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c [hunk 2]
```diff
@@ -82,9 +86,9 @@ static inline void _set_pktsnip(gnrc_pktsnip_t *pkt, gnrc_pktsnip_t *next,
 void gnrc_pktbuf_init(void)
 {
     mutex_lock(&_mutex);
-    _first_unused = (_unused_t *)_pktbuf;
+    _first_unused = (_unused_t *)_pktbuf_buf;
```
Comment (maribu) [inline sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:89]: This is where the alignment previously was not handled correctly. It was just casting a char array to an `unused_t` pointer.

Comment (miri64) [reply sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:89]: As `_pktbuf` was in static memory, this seems to be to me more of a false positive. But if it makes the compiler happier, I'm not blocking.

Comment (maribu) [reply sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c:89]: ### `test.c`
```C
#include <stdio.h>

static char a[3];
static char b[3];
static char c[3];

int main(void)
{
    printf("address of a = %p\n", (void *)a);
    printf("address of b = %p\n", (void *)b);
    printf("address of c = %p\n", (void *)c);

    return 0;
}
```

```
$ gcc -Wall -Wextra -pedantic -std=c99 -Os -o test test.c 
$ ./test
address of a = 0x5653e3786076
address of b = 0x5653e3786073
address of c = 0x5653e3786070
```

**Review by miri64 — APPROVED** [review]
Sadly, the text size changes (only tested on Cortex-M4 [slight increase] and MSP430 [slight decrease]), but the unittests for `gnrc_pktbuf_static` still pass with this PR.

```
[mlenders@sofia RIOT]<3 RIOT_CI_BUILD=1 BOARD=samr21-xpro make -C tests/unittests clean tests-pktbuf flash -j16 test
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/unittests'
Building application "tests_unittests" for "samr21-xpro" with MCU "samd21".

[INFO] edbg binary not found - building it from source now
CC= CFLAGS= make -C /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/edbg
[INFO] updating edbg /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/edbg/bin/.pkg-state.git-downloaded
echo 99d15460fcff723f73b16c29c8ca14bff4b33b20 > /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/edbg/bin/.pkg-state.git-downloaded
[INFO] patch edbg
[INFO] edbg binary successfully built!
   text	   data	    bss	    dec	    hex	filename
  25792	    132	   9312	  35236	   89a4	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/unittests/bin/samr21-xpro/tests_unittests.elf
/home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/edbg/edbg  --target samr21 --verbose --file /home/mlenders/Repositories/RIOT-OS/RIOT/tests/unittests/bin/samr21-xpro/tests_unittests.bin --verify || /home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/edbg/edbg  --target samr21 --verbose --file /home/mlenders/Repositories/RIOT-OS/RIOT/tests/unittests/bin/samr21-xpro/tests_unittests.bin --verify --program
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800004989 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification...
at address 0x4 expected 0xf1, read 0x49
Error: verification failed
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800004989 01.1A.00FB (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Programming......................................................................................................... done.
Verification......................................................................................................... done.
r
make[1]: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/unittests'
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
/home/mlenders/Repositories/RIOT-OS/RIOT/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Twisted not available, please install it if you want to use pyterm's JSON capabilities
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
.............................................
OK (45 tests)

make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/unittests'
```

ACK

**Comment by maribu** [comment]
Hmm, might be worth to take a look on why `.text` increases. I'll do so now.

**Comment by maribu** [comment]
Did you use `RIOT_VERSION=test`, btw?

**Comment by miri64** [comment]
> Did you use `RIOT_VERSION=test`, btw?

as dumped: I used `RIOT_CI_BUILD=1` which sets `RIOT_VERSION=buildtest` ;-)

**Comment by maribu** [comment]
For me `.text` gets smaller by 8 bytes in symbol `gnrc_pktbuf_init` for `samr21-xpro` with this PR (tested with `examples/gnrc_minimal`). Interestingly, the compiler previously used 4 byte-wise instructions for `_first_unused->next = NULL;` and again for byte-wise instructions for `_first_unused->size = sizeof(_pktbuf);`.

Apparently the compiler was aware that `_frist_unused` can indeed be unaligned and just created correct code for this.

**Comment by maribu** [comment]
> as dumped: I used `RIOT_CI_BUILD=1` which sets `RIOT_VERSION=buildtest` ;-)

Thanks for pointing this out. This is quite useful :-)

**Comment by miri64** [comment]
Then let's go :-)

**Comment by miri64** [comment]
> For me `.text` gets smaller by 8 bytes in symbol `gnrc_pktbuf_init` for `samr21-xpro` with this PR (tested with `examples/gnrc_minimal`). Interestingly, the compiler previously used 4 byte-wise instructions for `_first_unused->next = NULL;` and again for byte-wise instructions for `_first_unused->size = sizeof(_pktbuf);`.

While I tested with `samr21-xpro` I did not compare sizes for it. Just `nrf52dk`

**Comment by maribu** [comment]
Thanks :-)

**Pull request merged**